nativelib/src/main/resources/gc/immix/Heap.c

```diff
@@ -0,0 +1,253 @@
+
+    Heap_collect(heap, stack);
+
+    Object *object = (Object *)Allocator_alloc(heap->allocator, size);
+    if (object != NULL) {
+        ObjectHeader *objectHeader = &object->header;
+
+        Object_setObjectType(objectHeader, object_standard);
+        Object_setSize(objectHeader, size);
+        Object_setAllocated(objectHeader);
+    }
+
+    if (object == NULL) {
+        Heap_grow(heap, size);
+
+        object = (Object *)Allocator_alloc(heap->allocator, size);
+        assert(object != NULL);
+
+        ObjectHeader *objectHeader = &object->header;
+
+        Object_setObjectType(objectHeader, object_standard);
+        Object_setSize(objectHeader, size);
+        Object_setAllocated(objectHeader);
+    }
+
+    return Object_toMutatorAddress(object);
+}
+
+INLINE word_t *Heap_allocSmall(Heap *heap, uint32_t objectSize) {
+    // Add header
+    uint32_t size = objectSize + OBJECT_HEADER_SIZE;
+
+    assert(objectSize % WORD_SIZE == 0);
+    assert(size < MIN_BLOCK_SIZE);
+
+    Object *object = (Object *)Allocator_alloc(heap->allocator, size);
+    if (object != NULL) {
+        ObjectHeader *objectHeader = &object->header;
+        Object_setObjectType(objectHeader, object_standard);
+        Object_setSize(objectHeader, size);
+        Object_setAllocated(objectHeader);
+
+        return Object_toMutatorAddress(object);
+    } else {
+        return allocSmallSlow(heap, size);
+    }
+}
+
+word_t *Heap_alloc(Heap *heap, uint32_t objectSize) {
+    assert(objectSize % WORD_SIZE == 0);
+
+    if (objectSize + OBJECT_HEADER_SIZE >= LARGE_BLOCK_SIZE) {
+        return Heap_allocLarge(heap, objectSize);
+    } else {
+        return Heap_allocSmall(heap, objectSize);
+    }
+}
+
+void Heap_collect(Heap *heap, Stack *stack) {
+#ifdef DEBUG_PRINT
+    printf("\nCollect\n");
+    fflush(stdout);
+#endif
+    Marker_markRoots(heap, stack);
+    Heap_recycle(heap);
+
+#ifdef DEBUG_PRINT
+    printf("End collect\n");
+    fflush(stdout);
+#endif
+}
+
+void Heap_recycle(Heap *heap) {
+    BlockList_clear(&heap->allocator->recycledBlocks);
+    BlockList_clear(&heap->allocator->freeBlocks);
+
+    heap->allocator->freeBlockCount = 0;
+    heap->allocator->recycledBlockCount = 0;
+
+    heap->allocator->freeMemoryAfterCollection = 0;
+
+    word_t *current = heap->heapStart;
+    while (current != heap->heapEnd) {
+        BlockHeader *blockHeader = (BlockHeader *)current;
+        Block_recycle(heap->allocator, blockHeader);
+        // block_print(blockHeader);
+        current += WORDS_IN_BLOCK;
+    }
+    LargeAllocator_sweep(heap->largeAllocator);
+
+    if (!Allocator_canInitCursors(heap->allocator) ||
+        Allocator_shouldGrow(heap->allocator)) {
+        size_t increment = heap->smallHeapSize / WORD_SIZE * GROWTH_RATE / 100;
+        increment =
+            (increment - 1 + WORDS_IN_BLOCK) / WORDS_IN_BLOCK * WORDS_IN_BLOCK;
+        Heap_grow(heap, increment);
+    }
+    Allocator_initCursors(heap->allocator);
+}
```

Comment:
We need to gracefully fail if we reach the limit of available heap memory, e.g. throw OutOfMemoryError.